[JIT] linear scan memory planning strategy

makslevental opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63873
* #64350
* #64349
* __->__ #64348
* #64347
* #62995

Linear scan (a heuristic based on http://web.cs.ucla.edu/~palsberg/course/cs132/linearscan.pdf) memory planning strategy.

Differential Revision: [D30769099](https://our.internmc.facebook.com/intern/diff/D30769099)

Commit:
[JIT] linear scan memory planning strategy

Linear scan (a heuristic based on https://www.usenix.org/legacy/events/vee05/full_papers/p132-wimmer.pdf) memory planning strategy.

[ghstack-poisoned]

Commit:
Update on "[JIT] linear scan memory planning strategy"


Linear scan (a heuristic based on https://www.usenix.org/legacy/events/vee05/full_papers/p132-wimmer.pdf) memory planning strategy.

[ghstack-poisoned]

makslevental committed:
Update on "[JIT] linear scan memory planning strategy"


Linear scan (a heuristic based on https://www.usenix.org/legacy/events/vee05/full_papers/p132-wimmer.pdf) memory planning strategy.

[ghstack-poisoned]

Commit:
Update on "[JIT] linear scan memory planning strategy"


Linear scan (a heuristic based on https://www.usenix.org/legacy/events/vee05/full_papers/p132-wimmer.pdf) memory planning strategy.

[ghstack-poisoned]

makslevental committed:
Update on "[JIT] linear scan memory planning strategy"


Linear scan (a heuristic based on https://www.usenix.org/legacy/events/vee05/full_papers/p132-wimmer.pdf) memory planning strategy.

[ghstack-poisoned]

Commit:
Update on "[JIT] linear scan memory planning strategy"


Linear scan (a heuristic based on https://www.usenix.org/legacy/events/vee05/full_papers/p132-wimmer.pdf) memory planning strategy.

[ghstack-poisoned]

Commit:
Update on "[JIT] linear scan memory planning strategy"


Linear scan (a heuristic based on https://www.usenix.org/legacy/events/vee05/full_papers/p132-wimmer.pdf) memory planning strategy.

[ghstack-poisoned]

Commit:
Update on "[JIT] linear scan memory planning strategy"


Linear scan (a heuristic based on https://www.usenix.org/legacy/events/vee05/full_papers/p132-wimmer.pdf) memory planning strategy.

[ghstack-poisoned]

Commit:
Update on "[JIT] linear scan memory planning strategy"


Linear scan (a heuristic based on https://www.usenix.org/legacy/events/vee05/full_papers/p132-wimmer.pdf) memory planning strategy.

[ghstack-poisoned]

makslevental commented:
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30769099).

Commit:
Update on "[JIT] linear scan memory planning strategy"


Linear scan (a heuristic based on https://www.usenix.org/legacy/events/vee05/full_papers/p132-wimmer.pdf) memory planning strategy.

Differential Revision: [D30769099](https://our.internmc.facebook.com/intern/diff/D30769099)

[ghstack-poisoned]

Commit:
Update on "[JIT] linear scan memory planning strategy"


Linear scan (a heuristic based on https://www.usenix.org/legacy/events/vee05/full_papers/p132-wimmer.pdf) memory planning strategy.

Differential Revision: [D30769099](https://our.internmc.facebook.com/intern/diff/D30769099)

[ghstack-poisoned]

Commit:
Update on "[JIT] linear scan memory planning strategy"


Linear scan (a heuristic based on https://www.usenix.org/legacy/events/vee05/full_papers/p132-wimmer.pdf) memory planning strategy.

Differential Revision: [D30769099](https://our.internmc.facebook.com/intern/diff/D30769099)

[ghstack-poisoned]

Commit:
Update on "[JIT] linear scan memory planning strategy"


Linear scan (a heuristic based on https://www.usenix.org/legacy/events/vee05/full_papers/p132-wimmer.pdf) memory planning strategy.

Differential Revision: [D30769099](https://our.internmc.facebook.com/intern/diff/D30769099)

[ghstack-poisoned]

Commit:
Update on "[JIT] linear scan memory planning strategy"


Linear scan (a heuristic based on https://www.usenix.org/legacy/events/vee05/full_papers/p132-wimmer.pdf) memory planning strategy.

Differential Revision: [D30769099](https://our.internmc.facebook.com/intern/diff/D30769099)

[ghstack-poisoned]

makslevental committed:
Update on "[JIT] linear scan memory planning strategy"


Linear scan (a heuristic based on https://www.usenix.org/legacy/events/vee05/full_papers/p132-wimmer.pdf) memory planning strategy.

Differential Revision: [D30769099](https://our.internmc.facebook.com/intern/diff/D30769099)

[ghstack-poisoned]

makslevental committed:
Update on "[JIT] linear scan memory planning strategy"


Linear scan (a heuristic based on https://www.usenix.org/legacy/events/vee05/full_papers/p132-wimmer.pdf) memory planning strategy.

Differential Revision: [D30769099](https://our.internmc.facebook.com/intern/diff/D30769099)

[ghstack-poisoned]

makslevental committed:
Update on "[JIT] linear scan memory planning strategy"


Linear scan (a heuristic based on https://www.usenix.org/legacy/events/vee05/full_papers/p132-wimmer.pdf) memory planning strategy.

Differential Revision: [D30769099](https://our.internmc.facebook.com/intern/diff/D30769099)

[ghstack-poisoned]

makslevental committed:
Update on "[JIT] linear scan memory planning strategy"


Linear scan (a heuristic based on https://www.usenix.org/legacy/events/vee05/full_papers/p132-wimmer.pdf) memory planning strategy.

Differential Revision: [D30769099](https://our.internmc.facebook.com/intern/diff/D30769099)

[ghstack-poisoned]

makslevental committed:
Update on "[JIT] linear scan memory planning strategy"


Linear scan (a heuristic based on https://www.usenix.org/legacy/events/vee05/full_papers/p132-wimmer.pdf) memory planning strategy.

Differential Revision: [D30769099](https://our.internmc.facebook.com/intern/diff/D30769099)

[ghstack-poisoned]

Krovatkin commented on torch/csrc/jit/passes/memory_planning/linear_scan.cpp:
```diff
@@ -0,0 +1,147 @@
+      continue;
+    }
+
+    TORCH_INTERNAL_ASSERT(
```
lmao nuclear shelter proof 😸

makslevental replied:
is this the wrong thing? should i use something else?

Krovatkin commented on torch/csrc/jit/passes/memory_planning/linear_scan.cpp:
```diff
@@ -0,0 +1,147 @@
+      size_t candidate_size = candidate_reg.size;
+      active[curr_range] = {candidate_offset, aligned_curr_size};
+      // split region (potentially)
+      if (candidate_size > aligned_curr_size) {
```
looks like duplication of the logic above? Maybe we should refactor it into a helper?

makslevental replied:
it's true but it's mostly duplication of unpacking structs and such. only duplicated logic is the check which is more intelligible in context i think. not sure - could go either way (we should flip a coin)

Krovatkin commented on torch/csrc/jit/passes/memory_planning/linear_scan.cpp:
```diff
@@ -0,0 +1,147 @@
+    curr_end_offset += aligned_curr_size;
+  }
+
+  // expire any remaining intervals
```
I feel like this comment is very helpful, but I struggle to come up with something better. The only side effect you care about here is moving remaining intervals to `allocated_ranges` .

makslevental replied:
discussed offline

makslevental replied:
i think you meant *not very helpful? agreed. will add a couple more words

Krovatkin commented on torch/csrc/jit/passes/memory_planning/linear_scan.cpp:
```diff
@@ -0,0 +1,147 @@
+  std::vector<MemAllocation> allocations;
+  allocations.reserve(allocated_ranges.size());
+  for (const auto& item : allocated_ranges) {
+    allocations.push_back({item.first, item.second});
```
err this looks fishy... you already reserved enough elements for `allocated_ranges` but then you are pushing back those on top reserved ones?

makslevental replied:
discussed offline

Krovatkin commented on torch/csrc/jit/passes/memory_planning/linear_scan.h:
```diff
@@ -0,0 +1,16 @@
+namespace torch {
+namespace jit {
+
+using EndSortedLiveRangeMap =
```
it's only used in the `cpp` file, why don't we move it there?

makslevental replied:
i can't remember if it's used in any of the other diffs but it's a vaguely useful typedef? not sure. another coin flip.

Krovatkin commented on test/cpp/jit/test_memory_planning.cpp:
```diff
@@ -432,5 +432,32 @@ TEST(MemoryPlannerTest, LSTMNaive) {
+      {448, 448, TTP(((Vec{10, 10})), ((Vec{10, 1})))},
+      {448, 896, TTP(((Vec{10, 10})), ((Vec{10, 1})))},
+  };
+  testSmall(expected_storage, expected_allocs, Strategy::LINEAR_SCAN);
```
I take it all strategies are tested with only these two?

makslevental replied:
correct. longer models would've inflated the verification spec data too much i think

Krovatkin reviewed (approved):
:shipit: Could you please take a look at https://github.com/pytorch/pytorch/pull/64348#discussion_r714344127 before landing this?

makslevental commented on torch/csrc/jit/passes/memory_planning/linear_scan.cpp:
```diff
@@ -0,0 +1,147 @@
+  SortedLiveRangeMap<MemRegion> allocated_ranges;
+
+  size_t curr_end_offset = 0;
+  auto allocate_inactive_ranges = [&](UniqueLiveRange curr_range) {
```
agreed. will do

makslevental commented on torch/csrc/jit/passes/memory_planning/linear_scan.cpp:
```diff
@@ -0,0 +1,147 @@
+    // find the "right" region; in order of preference:
+    // 1. tightest fit free region i.e. smallest i.e. first match since
+    // avail_regions is sorted by size
+    // 2. swap with latest ending active live range that is big enough (spilling
```
i mean it probably doesn't if the data is indicative but this is the heuristic from the original linear scan paper


![Screenshot from 2021-09-22 17-20-11](https://user-images.githubusercontent.com/5657668/134429674-a2330ecc-c110-43da-99d7-dae639204161.png)

makslevental committed:
Update on "[JIT] linear scan memory planning strategy"


Linear scan (a heuristic based on http://web.cs.ucla.edu/~palsberg/course/cs132/linearscan.pdf) memory planning strategy.

Differential Revision: [D30769099](https://our.internmc.facebook.com/intern/diff/D30769099)

[ghstack-poisoned]

makslevental commented:
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30769099).
